testgrid/conformance/upload_e2e.py

```diff
@@ -0,0 +1,206 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script parses conformance test output to produce testgrid entries
+#
+# Assumptions:
+# - there is one log file and one JUnit file (true for current conformance tests..)
+# - the log file contains ginkgo's output (true for kubetest and sonobuoy..)
+# - the ginkgo output will give us start / end time, and overall success
+#
+# - the start timestamp is suitable as a testgrid ID (unique, monotonic)
+#
+# - the test ran in the current year unless --year is provided
+# - the timestamps are parsed on a machine with the same local time (zone) 
+#   settings as the machine that produced the logs
+#
+# The log file is the source of truth for metadata, the JUnit will be consumed
+# by testgrid / gubernator for individual test case results
+#
+# Usage: see README.md
+
+
+import re
+import sys
+import time
+from datetime import datetime
+import argparse
+import json
+import subprocess
+from os import path
+
+# NOTE e2e logs use go's time.StampMilli ("Jan _2 15:04:05.000")
+# Example log line with a timestamp:
+# Jan 26 06:38:46.284: INFO: Running AfterSuite actions on all node
+# the third ':' separates the date from the rest
+E2E_LOG_TIMESTAMP_RE = re.compile('(... .\d \d\d:\d\d:\d\d\.\d\d\d):.*')
+
+# Ginkgo gives a line like the following at the end of successful runs:
+# SUCCESS! -- 123 Passed | 0 Failed | 0 Pending | 587 Skipped PASS
+# we match this to detect overall success
+E2E_LOG_SUCCESS_RE = re.compile('SUCCESS! -- .* PASS')
+
+def parse_e2e_log_line_timestamp(line, year):
+    """parses a ginkgo e2e log line for the leading timestamp
+
+    Args:
+        line (str) - the log line
+        year (str) - 'YYYY'
+
+    Returns:
+        timestamp (datetime.datetime) or None
+    """
+    match = E2E_LOG_TIMESTAMP_RE.match(line)
+    if match is None:
+        return None
+    # note we add year to the timestamp because the actual timestamp doesn't
+    # contain one and we want a datetime object...
+    timestamp = year+' '+match.group(1)
+    return datetime.strptime(timestamp, '%Y %b %d %H:%M:%S.%f')
+
+def parse_e2e_logfile(f, year):
+    """parse e2e logfile at path, assuming the log is from year
+    
+    Args:
+        f (file): the log file, iterated for lines
+        year (str): YYYY year logfile is from
+
+    Returns:
+        started (datetime.datetime), finished (datetime.datetime), passed (boolean)
+    """
+    started = finished = None
+    passed = False
+    for line in f:
+        # try to get a timestamp from each line, keep the first one as
+        # start time, and the last one as finish time
+        ts = parse_e2e_log_line_timestamp(line, year)
+        if started is None:
```

Comment:
Use implicit comparision whenever possible (`if not started`)

https://google.github.io/styleguide/pyguide.html?showone=True/False_evaluations#True/False_evaluations

Reply:
done.